lib/livebook/runtime/node_pool.ex

```diff
@@ -0,0 +1,84 @@
+defmodule Livebook.Runtime.NodePool do
+  use GenServer, async: false
+
+  # A pool for reusing child node names.
+  #
+  # `pool` refers to the list of unused names
+  #
+  # `buffer_time` refers to time the pool waits before
+  # adding a name to `pool`,  which by default is 1 minute.
+
+  @default_time 6_000
+
+  # Client interface
+
+  @doc """
+  Starts the node pool with the buffer time as 1 minute.
+  """
+  def start do
+    GenServer.start(__MODULE__, @default_time, name: __MODULE__)
+  end
+
+  @doc """
+  Starts the node pool with the buffer time as input.
+  """
+  def start(buffer_time) do
+    GenServer.start(__MODULE__, buffer_time, name: __MODULE__)
+  end
+
+  @doc """
+  Retuns a node name.
+
+  Generates a new one if node is empty, or takes on from pool.
+  """
+  def get_name(parent) do
+    GenServer.call(__MODULE__, {:get_name, parent})
+  end
+
+  # Server side code
+
+  @impl GenServer
+  def init(buffer_time) do
+    :net_kernel.monitor_nodes(true)
+    {:ok, %{pool: [], buffer_time: buffer_time}}
+  end
+
+  @impl GenServer
+  def handle_call({:get_name, parent}, _, state) do
+    {:reply, name(state, parent), state}
```

Comment:
I think we can have a race condition here: two processes can call `get_name` at the same time and both of them will receive the same name and only one of them will in.

I think once we call `get_name`, we need to remove it from the pool. There is a chance there is a crash and we will lose the node name forever... but I think that's fine. Then you can ignore nodeup messages. :)